### PR TITLE
API/REF: Clone before fitting in Incremental.fit

### PR DESCRIPTION
**Pull request by TomAugspurger**
This changes Incremental.fit to clone before fitting.

Alternative to https://github.com/dask/dask-ml/pull/207

### CODE REVIEW

**Comment by TomAugspurger** [comment]
cc @stsievert

File: tests/test_incremental.py [hunk 1]
```diff
@@ -26,11 +27,12 @@ def test_incremental_basic(scheduler, xy_classification):
 
         assert result is clf
 
-        assert isinstance(result.estimator.coef_, np.ndarray)
-        np.testing.assert_array_almost_equal(result.estimator.coef_,
+        assert isinstance(result.estimator_.coef_, np.ndarray)
```
Comment (TomAugspurger) [inline tests/test_incremental.py]: This is the API change.

File: tests/test_incremental.py [hunk 2]
```diff
@@ -96,22 +99,20 @@ def test_scoring_string(scheduler, xy_classification, scoring):
         assert callable(check_scoring(clf, scoring=scoring))
         clf.fit(X, y, classes=np.unique(y))
         clf.score(X, y)
-        clf.estimator.score(X, y)
```
Comment (TomAugspurger) [inline tests/test_incremental.py]: I don't think this was doing what we wanted earlier.

When Incremental is created with `scoring`, `clf.score(X, y)` will in general be different then `clf.estimator_.score(X, y)`. The latter always uses the "default" scoring for that estimator (e.g. accuracy score). I opted to just remove this line.

File: tests/test_incremental.py [hunk 3]
```diff
@@ -96,22 +99,20 @@ def test_scoring_string(scheduler, xy_classification, scoring):
...
 
 
 def test_fit_ndarrays():
     X = np.ones((10, 5))
-    y = np.ones(10)
+    y = np.concatenate([np.zeros(5), np.ones(5)])
```
Comment (TomAugspurger) [inline tests/test_incremental.py]: Needed a mix of zeros and ones. I'm not entirely sure what changed to necessitate that...

File: tests/test_incremental.py [hunk 4]
```diff
@@ -122,7 +123,7 @@ def test_score_ndarrays():
     inc = Incremental(sgd, scoring='accuracy')
 
     inc.partial_fit(X, y, classes=[0, 1])
-    inc.fit(X, y)
+    inc.fit(X, y, classes=[0, 1])
```
Comment (TomAugspurger) [inline tests/test_incremental.py]: need this, as `classes_` no longer leasks through the `partial_fit` to the `fit` (a good thing!)

**Comment by stsievert** [comment]
I think if we implement `fit` support for `Incremental` we should loop over the data `max_iter` times beforehand.

For more complete thoughts on the API, see https://github.com/dask/dask-ml/pull/207#issuecomment-401834441

**Comment by TomAugspurger** [comment]
Agreed that having a `max_iter` parameter would be useful for Incremental. However, exposes a difficulty alluded to in https://github.com/dask/dask-ml/issues/200 and how `Incremental.get/set_params` works. It would be ambiguos whether `max_iter` should go to `Incremental.max_iter` or `Incremental.estimator.max_iter`.

The solution is to probably to have `Incremental` to specify a name when creating the `Incremental`.

See https://stackoverflow.com/a/35534452/1889400 for how `VotingClassifier` works.

So either

```python
inc = Incremental(SGDClassifier(), name='sgdclassifier')
inc = Incremental(('sgdclassifier', SGDClassifier())
```

In the latter case, we would have a `make_incremental` helper that works like `make_pipeline`.

Then in a context where you need to get / set params, it'd be

```python
param_grid = {
    'sgdclassifier__alpha': [0.1, 10.0],
    'sgdclassifier__max_iter': [1000, 2000],
    'max_iter': [1, 5],
}

clf = GridSearchCV(inc, param_grid)
```

**Comment by TomAugspurger** [comment]
Looks like https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/metaestimators.py has some utilities for get / set params in a meta-estimator.

**Comment by TomAugspurger** [comment]
55fccf20e0c2204f5149db7161047ff31184ba85 has the changes.

Turns out BaseEstimator already does the right thing, we don't have to mess with `_BaseComposition`.

I'm encouraged by the lines of code we could delete from `Incremental`.

For now, I haven't made the name configurable. If you want to set the parameters for the underlying estimator, then you use `estimator__<parameter>`. We can make the `estimator` part configurable in the future if there's a demand for it.

**Comment by TomAugspurger** [comment]
I'll implement `max_iter` now.

@stsievert, do you have a feeling for what a good default would be? I'm leaning towards `max_iter=1`, but am not sure.

**Comment by mrocklin** [comment]
What does Scikit-Learn do in situations like these?  Do they have some sort
of convergence criterion that they check?  Perhaps we can do the same?

**Comment by TomAugspurger** [comment]
It varies from model to model.

https://github.com/scikit-learn/scikit-learn/issues/1626 is the general API that we would want to follow here, when that is implemented.

**Comment by TomAugspurger** [comment]
A basic version of `max_iter` should be working now.

**Comment by TomAugspurger** [comment]
I'm finding this diff somewhat difficult to digest. I'm going to split off the `max_iter` to a second PR.

**Comment by stsievert** [comment]
I'm on board with this now. I think `Incremental` should implement `fit` and have a  `max_iter` param.

> do you have a feeling for what a good default would be? I'm leaning towards max_iter=1, but am not sure.

I think I would default to `max_iter=5` – that's what sklearn's SGDClassifier `max_iter` defaults to, at least when no stopping criterion set in <v0.21. They set some stopping criteria in >= v0.21 and increase `max_iter`. For simple problems I'd say that 5 epochs is good enough.

This *should* depend on model complexity but I'm okay leaving that. e.g., fitting a linear model with twice as many features will require about `sqrt(2)` more samples to be seen by the optimization algorithm.
